### PR TITLE
Support joint export with unused parameter grads

### LINKED ISSUE #101192: AOTAutograd export path does not support training graphs with parameters that do not receive gradients.
See the comment [here](https://github.com/pytorch/pytorch/pull/100587/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416R3786): today, the AOTAutograd export path expects that every parameter in the module you export will receive some sort of gradient. It's possible to create a case where this doesn't happen. We can support it, but I'd like to gauge how much request there is for this first, since it's more of an edge case.

Example:

```
def f(p1, p2):
    # p1 will never receive any gradients. p2 will though
    return p1.detach() * p2
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186542

AOTAutograd's flattened joint export path assumed that every grad-requiring forward input would produce a non-None gradient. When a parameter participated only through a detached value, autograd correctly returned None for that input, but export raised an internal assertion instead of representing the remaining real gradients.

Record the flattened input indices that actually produce gradient tensors, omit None gradients from the joint graph outputs, and build the backward signature from those recorded indices. This preserves the existing mapping for normal joint exports while allowing parameters that legitimately do not receive gradients.

The alternative would be to materialize missing gradients as zeros, but that would incorrectly advertise a gradient output for parameters that autograd reports as unused. Omitting those outputs matches the existing backward signature contract, which maps graph outputs only to emitted gradients.

Fixes #101192

Generated by my agent

Test Plan:

- Reproducer using aot_export_module(..., trace_joint=True, output_loss_index=0) on a module returning ((p1.detach() * p2).sum(),): passed after fix and produced a backward signature only for p2.

- python test/functorch/test_aotdispatch.py -k test_aot_export_module_joint_unused_parameter_gradient: blocked before collection by unrelated local torchvision mismatch (operator torchvision::nms does not exist).

- Python runpy import-hook variant of the same targeted test with torchvision disabled: passed, Ran 1 test, OK.

- Python runpy import-hook variant with -k test_aot_export_module_joint: passed, Ran 2 tests, OK.

- git diff --check

- lintrunner -a

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the Claude review feedback in the latest update: removed the duplicate comment, clarified the parameter/buffer count naming, added a regression assertion for the loss output, and documented the token/signature layout fix in the commit message. While doing that review follow-up, the fresh local review also caught that joint graph BackwardSignature.loss_output needed the same output-layout offset; that is fixed and covered by the updated joint export test expectation.

File: torch/_functorch/_aot_autograd/input_output_analysis.py [hunk 1]
```diff
@@ -455,40 +455,56 @@ def create_graph_signature(
...
-        gradients_to_parameters = {
-            backward_output_names[next(grad_index)]: param_names[i]
-            for i, param in enumerate(params_and_buffers_flat)
-            if param.requires_grad
-        }
-
-        gradients_to_user_inputs = {
-            backward_output_names[next(grad_index)]: graph_input_names[
-                i + len(params_and_buffers_flat)
+        if traced_gradient_input_indices is None:
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/input_output_analysis.py:452]: [cleanup] This fallback is unreachable with the current callers: the only `trace_joint=True` caller (`aot_autograd.py`) always passes `traced_gradient_input_indices`, and `torch/export/_trace.py` calls with `trace_joint=False`. Beyond being dead, it re-encodes the exact assumption this PR removes -- it derives the indices from `requires_grad` alone -- so any future `trace_joint=True` caller relying on the default would hit the `len(traced_gradient_input_indices) != len(backward_output_names)` assertion again for an unused-grad module, i.e. silently reintroduce #101192. Consider dropping the default and making `traced_gradient_input_indices` a required keyword argument so the contract is explicit.

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] The regression test covers only the degenerate omission case (`p1.detach() * p2`: two parameters, no input mutations, no user inputs), but this PR enables two further new paths that previously raised "Found a parameter that did not receive a gradient" and now silently build a backward signature with nothing exercising them:

1. A user *input* that requires grad but receives no gradient -- the `gradients_to_user_inputs` branch in `create_graph_signature`.
2. An unused gradient combined with input mutations -- where the rewritten `num_tokens + num_mutated_inp_runtime_indices` output-offset arithmetic is load-bearing. The existing batchnorm test has mutations but no omission; the new test has omission but no mutations, so the combination is unexercised.

These are the realistic-usage shapes the issue motivates (MoE blocks have many unused experts alongside normalization buffers and activation inputs), not edge cases. A wrong index here does not crash -- it maps a graph output to the wrong parameter/input or names the wrong `loss_output`, surfacing as incorrect gradients downstream. Tracing the mappings suggests they are correct (the mutation offset is on the output side and omission is on the gradient side, so the two are orthogonal), but a silent-correctness path is exactly what a test must backstop. Please add coverage: a module that mutates a buffer (or returns a detached non-loss output) alongside an unused parameter, plus one with an unused grad-requiring user input.

File: torch/_functorch/_aot_autograd/input_output_analysis.py (hunk 1 above)
Comment (jansel) [reply torch/_functorch/_aot_autograd/input_output_analysis.py:452]: My agent says I removed that fallback and made the contract explicit: trace_joint=True now requires traced_gradient_input_indices from the actual joint trace, while the non-joint torch/export caller passes None explicitly. This avoids reconstructing gradient outputs from requires_grad and prevents future callers from reintroducing the unused-gradient assumption.

**Comment by jansel** [comment]
My agent says I addressed the requested coverage in the latest update: added regression tests for an unused grad-requiring user input and for an unused parameter gradient combined with buffer mutation/output offsets. I also tightened the assertions to check exact gradient-output mappings, removed the trace_joint fallback that reconstructed indices from requires_grad, and reran the targeted joint export tests plus lintrunner.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
